store/tikv/region_cache.go

```diff
@@ -70,76 +70,83 @@ type Region struct {
-	stores         []*Store // stores in this region
-	storeFails     []uint32 // snapshots of store's fail, need reload when `storeFails[curr] != stores[cur].fail`
+	currIdx [EngineCnt]int32
+	stores  [EngineCnt][]PeerStore // stores in this region
+}
+
+// PeerStore represent a store peer in special engine store list.
+type PeerStore struct {
+	Idx       int          // Index at PeerStore list for special engine.
+	Peer      *metapb.Peer // Store's peer info
+	Store     *Store       // Store info
+	storeFail uint32       // snapshots of store's fail, need reload when `peerStore.storeFail != stores[cur].fail`
 }
 
 // clone clones region store struct.
 func (r *RegionStore) clone() *RegionStore {
-	storeFails := make([]uint32, len(r.stores))
-	for i, e := range r.storeFails {
-		storeFails[i] = e
+	nr := &RegionStore{
+		currIdx: r.currIdx,
 	}
-	return &RegionStore{
-		workTiFlashIdx: r.workTiFlashIdx,
-		workTiKVIdx:    r.workTiKVIdx,
-		stores:         r.stores,
-		storeFails:     storeFails,
+	for engineType := range r.stores {
+		nr.stores[engineType] = r.stores[engineType]
 	}
+	return nr
 }
 
 // return next follower store's index
-func (r *RegionStore) follower(seed uint32) int32 {
-	l := uint32(len(r.stores))
+func follower(candidates []PeerStore, idx *int32, seed uint32) int32 {
+	l := uint32(len(candidates))
 	if l <= 1 {
-		return r.workTiKVIdx
+		return *idx
 	}
-
 	for retry := l - 1; retry > 0; retry-- {
 		followerIdx := int32(seed % (l - 1))
-		if followerIdx >= r.workTiKVIdx {
+		if followerIdx >= *idx {
 			followerIdx++
 		}
-		if r.stores[followerIdx].storeType != kv.TiKV {
-			continue
-		}
-		if r.storeFails[followerIdx] == atomic.LoadUint32(&r.stores[followerIdx].fail) {
-			return followerIdx
+		c := candidates[followerIdx]
+		if c.storeFail == atomic.LoadUint32(&c.Store.fail) {
+			return int32(c.Idx)
 		}
 		seed++
 	}
-	return r.workTiKVIdx
+	return *idx
 }
 
 // init initializes region after constructed.
-func (r *Region) init(c *RegionCache) {
+func (r *Region) init(bo *Backoffer, c *RegionCache) error {
 	// region store pull used store from global store map
 	// to avoid acquire storeMu in later access.
-	rs := &RegionStore{
-		workTiKVIdx:    0,
-		workTiFlashIdx: 0,
-		stores:         make([]*Store, 0, len(r.meta.Peers)),
-		storeFails:     make([]uint32, 0, len(r.meta.Peers)),
-	}
+	var idx [EngineCnt]int
+	rs := &RegionStore{}
 	for _, p := range r.meta.Peers {
 		c.storeMu.RLock()
 		store, exists := c.storeMu.stores[p.StoreId]
 		c.storeMu.RUnlock()
 		if !exists {
 			store = c.getStoreByStoreID(p.StoreId)
 		}
-		rs.stores = append(rs.stores, store)
-		rs.storeFails = append(rs.storeFails, atomic.LoadUint32(&store.fail))
+		_, err := store.initResolve(bo, c)
+		if err != nil {
+			return err
+		}
+		i := idx[store.storeType]
+		idx[store.storeType] = i + 1
+		rs.stores[store.storeType] = append(rs.stores[store.storeType], PeerStore{
```

Comment:
Seems `var idx [EngineCnt] int` is not necessary? We can know the store's position after it's appended in `rs.stores`.